### PR TITLE
sys/arduino: Added Wire (I2C) interface

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR provides an implementation of the Arduino `Wire` library.

### Testing procedure

See PR #12518 

### Issues/PRs references

The changes in PR #7654 might require changes of the handling of the pseudomodule `arduino_lib`.

### CODE REVIEW

**Comment by smlng** [comment]
Cool! I'll try to have a closer look and test this soon(ish).

File: sys/arduino/wireport.cpp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    txBufferLength = 0;
+}
+
+uint8_t WirePort::endTransmission(uint8_t stop)
+{
+    DEBUG("[wire] %s: stop %d\n", __func__, stop);
+
+    uint8_t wrote = 0;
+    if (i2c_acquire(ARDUINO_I2C_DEV) == 0 &&
+        i2c_write_bytes(ARDUINO_I2C_DEV, txAddress, txBuffer, txBufferLength,
```
Comment (leandrolanzieri) [inline sys/arduino/wireport.cpp]: By checking the returned value of `i2c_write_bytes` you could know the specific error (i.e. no device responded on the address or no response after writing bytes) and return the appropriate error code.

File: sys/arduino/wireport.cpp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    txBufferLength = txBufferIndex;
+
+    return 1;
+}
+
+size_t WirePort::write(const uint8_t *data, size_t size)
+{
+    DEBUG("[wire] %s: data %p, size %d\n", __func__, data, size);
+    for (size_t i = 0; i < size; i++) {
+        if (!write(data[i])) {
```
Comment (leandrolanzieri) [inline sys/arduino/wireport.cpp]: Would it make sense to have a flag to indicate that the data didn't fit in the TX buffer? Right now the `endTransmission` function can't check that error case (return value 1).

File: sys/arduino/include/Arduino.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * @brief       Wrapper to keep source code compatibility for Arduino.h
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @file
+ * @{
+ */
+
+#ifndef ARDUINO_H
+#define ARDUINO_H
+
+#include "arduino.hpp"
```
Comment (leandrolanzieri) [inline sys/arduino/include/Arduino.h]: I think `#include "arduino.hpp"` should be moved inside the `#ifdef __cplusplus` block (but not inside the `extern "C"`).

**Comment by gschorcht** [comment]
Let's pickup the game again. I had an email conversation with someone who was asking whether it is possible to use Arduino driver libraries for I2C sensors in RIOT. Having the Wire interface library would be a step in this direction.

In addition to the wire library, we need a common approach to automatically import Arduino libraries as packages. I'll think about such an approach as soon as I have some time left.

**Comment by gschorcht** [comment]
@cladmi When I was trying to rebase this, I ran into conflicts since the make system has been changed with PR #7654. With this PR, I tried to define a pseudomodule arduino_lib so that it became possible not only to compile a complete sketch but also some Arduino libraries. Do have any idea, how I could do the same with the new make system as defined in PR #7654?

**Comment by cladmi** [comment]
Is the following the behavior your want?

If somebody wants the sketch, it does not use your lib.
If somebody wants your lib, it does not use the sketch.
Both sketch and lib depend on a "common" arduino part.

If this is the case, I think the current `arduino` module should be split between the sketch and the common that you also use.

So either rename it to `arduino_sketches`, and when an application that wants it, should say `USEMODULE += arduino_sketches`, and put the common as `arduino`.
Or it could be `arduino` == sketches, `arduino_lib` your thing, `arduino_common` common things required by both (if needed).

That depends on what is wanted to give as an interface to users.

Currently there was an automatic `arduino` is implemented by `arduino_sketches` that was done in the `sys/arduino/Makefile.include` and this goes against your need as it should be defined in `Makefile.dep` if keeping the current naming.

Depending on the naming wanted, I can do the change.

**Comment by gschorcht** [comment]
@cladmi Thanks for explaination. I will restructure the PR a bit. I will leave this PR for the implemenation of Wire (I2C) implentation and will prepare a new PR for the make system.

**Comment by gschorcht** [comment]
@cladmin The situation was the following:

You have a directory with an Arduino application, for example `examples/hello-world-arduino`. The `Makefile` enables the `arduino` module
```
USEMODULE += arduino
```
In this case, the make system expects a file with extension `.sketch` from which the file `_sketch.cpp` is generated and compiled afterwards.

However, when you have an Arduino library, the directory usually contains only a `.h` and a `.cpp` file, for example:
```
pkg/vl53l1x_arduino/VL53L1X.cpp
pkg/vl53l1x_arduino/VL53L1X.h
```
The compilation of such a directory requires same compiler configuration as for an Arduino application using `sys/arduino` library.

So the idea was just to declare a pseudomodule `arduino_lib` in addition to the `arduino` psdeudomodule. The makefile would then enable
```
USEMODULE += arduino_lib
```
which enables module `arduino` implicitly. The make system wouldn't try to generate `_sketch.cpp` from a `.sketch` file.

Therefore I used the following construct in former `sys/arduino/Makefile.include`
```diff
- $(shell $(RIOTTOOLS)/arduino/pre_build.sh $(SRCDIR) $(APPDIR) $(SKETCHES))
+ # run the Arduino pre-build script if Arduino module is not used as library
+ ifeq (, $(filter arduino_lib, $(USEMODULE)))
+     $(shell $(RIOTTOOLS)/arduino/pre_build.sh $(SRCDIR) $(APPDIR) $(SKETCHES))
+ endif
```
and in `sys/Makefile.dep`
```diff
+ ifneq (, $(filter arduino_lib, $(USEMODULE)))
+     USEMODULE += arduino
+ endif  
```

It would be really cool if it would be possible to import Arduino libraries as packages and to compile them since there are thousands of drivers available for Arduino IDE.

**Comment by maribu** [comment]
@gschorcht: You didn't have reacted to some open comments of @smlng yet. Otherwise, I'd test and (hopefully) ACK.

File: sys/arduino/include/wireport.hpp [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+     * a STOP condition after the request.
+     *
+     * Copied from https://www.arduino.cc/en/Reference/WireRequestFrom
+     *
+     * @param[in]   addr    7-bit address of the device to request bytes from
+     * @param[in]   size    Number of bytes to request
+     *
+     * @return  number of bytes returned from the slave device
+     */
+	uint8_t requestFrom(uint8_t addr, uint8_t size);
```
[the remark this answers is not in the document]
Comment (gschorcht) [reply sys/arduino/include/wireport.hpp]: Already fixed some time ago.

File: sys/arduino/include/wireport.hpp [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+     * in-between calls to #beginTransmission and #endTransmission.
+     *
+     * @param[in]   data    Array of data to send as bytes
+     * @param[in]   size    Number of bytes to transmit
+     *
+     * Copied from https://www.arduino.cc/en/Reference/WireWrite
+     *
+     * @return  number of bytes queued
+     */
+	virtual size_t write(const uint8_t *data, size_t size);
```
[the remark this answers is not in the document]
Comment (gschorcht) [reply sys/arduino/include/wireport.hpp]: Already fixed some time ago.

File: sys/arduino/include/wireport.hpp [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+     * @retval  3   received NACK on transmit of data
+     * @retval  4   other error
+     */
+    uint8_t endTransmission(uint8_t stop);
+
+    /**
+     * @brief   Request bytes from a I2C slave device
+     *
+     * Used by the master to request bytes from a slave device. The bytes may
+     * then be retrieved with the #available and #read methods. Sends always
```
[the remark this answers is not in the document]
Comment (gschorcht) [reply sys/arduino/include/wireport.hpp]: Unfortunatly, autolinking doesn't work for `read` and `ARDUINO_I2C_DEV` without the `#`. With the `#`, doxygen generate references as it should. `#` is a valid method to generate references to functions.

File: sys/arduino/wireport.cpp [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+#include "Wire.h"
+
+#include "log.h"
+#include "panic.h"
+#include "periph/i2c.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define WIRE_PORT_OK                    (0)
```
[the remark this answers is not in the document]
Comment (gschorcht) [reply sys/arduino/wireport.cpp]: I haven't use enum due to compatibility reasons. As declared in the copyright notice, the implementation is an adaptation of the original Arduino Wire library. These defines are from this original implementation. Therefore, I would prefer to leave it as it is.

**Comment by gschorcht** [comment]
@maribu Some of the comments were already taken into account. To some of them I gave an explaination.

**Comment by gschorcht** [comment]
@smlng Could you approve this PR? All your change request were addressed, either changed or you gave thumbs up for the comment.

**Comment by gschorcht** [comment]
@maribu The test application can be found in PR #12518.

File: sys/arduino/wireport.cpp [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+ * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ * Modified 2012 by Todd Krein (todd@krein.org) to implement repeated starts
+ * ```
+ *
+ * The documentation in this file is partially extracted from the original
+ * [Arduino Reference](https://www.arduino.cc/en/Reference/Wire) of this
+ * library which is published under the
+ * [Creative Commons Attribution-ShareAlike 3.0 License]
+ * (https://creativecommons.org/licenses/by-sa/3.0/).
```
[the remark this answers is not in the document]
Comment (gschorcht) [reply sys/arduino/wireport.cpp]: @smlng Fixed.

**Comment by smlng** [comment]
code-wise this looks good to me, can somebody tests this please?

**Comment by gschorcht** [comment]
@maribu Could you test it? If you don't have LIS3DH at hand, just tell me what other I2C device you have and I can prepare a test.

**Comment by maribu** [comment]
Sorry, not yet. I do not have a LIS3DH. I have BME280, BME680, SHT30, and [VL53L0X](https://github.com/adafruit/Adafruit_VL53L0X) at hands.

**Comment by gschorcht** [comment]
> Sorry, not yet. I do not have a LIS3DH. I have BME280, BME680, SHT30, and [VL53L0X](https://github.com/adafruit/Adafruit_VL53L0X) at hands.

I have attached a test application with an Arduino library for SHT31D, which also works for SHT30. I tested it with `arduino-mega2560`. The library is of poor quality but uses very intense I2C. Temperature and humidity are casted to `int` for output because of the lack of floating point outputs.

[tests_arduino_sht31d.zip](https://github.com/RIOT-OS/RIOT/files/3759287/tests_arduino_sht31d.zip)

**Comment by maribu** [comment]
Tested on an Arduino Nano:

Output of the terminal with incorrect I2C address:

```
2019-10-23 20:02:08,224 # main(): This is RIOT! (Version: 2020.01-devel-197-g1de8b-g-ard)
2019-10-23 20:02:08,258 # ClosedCube SHT3X-D Periodic Mode Example
2019-10-23 20:02:08,291 # supports SHT30-D, SHT31-D and SHT35-D
2019-10-23 20:02:08,292 # Serial #0
2019-10-23 20:02:08,454 # [ERROR] Cannot start periodic mode
2019-10-23 20:02:08,462 # Periodic Mode: [ERROR] Code #-40
2019-10-23 20:02:08,745 # Periodic Mode: [ERROR] Code #-40
2019-10-23 20:02:09,029 # Periodic Mode: [ERROR] Code #-40
...
```

Capture with logic analyzer (still incorrect I2C address):

Picture of I2C Capture

Output of the terminal with I2C address matching device:

```
2019-10-23 20:09:31,983 # main(): This is RIOT! (Version: 2020.01-devel-197-g1de8b-g-ardJ�(���Չ��SHT3X-D Periodic Mode Example
2019-10-23 20:09:32,017 # supports SHT30-D, SHT31-D and SHT35-D
2019-10-23 20:09:32,019 # Serial #28566
2019-10-23 20:09:32,150 # Periodic Mode: T=23C, RH=63%
2019-10-23 20:09:32,430 # Periodic Mode: T=23C, RH=63%
2019-10-23 20:09:32,710 # Periodic Mode: T=23C, RH=63%
```


So: Errors seem to be correctly reported to the libraries (first try with incorrect address) and with correct address it worked as advertised.

**Comment by maribu** [comment]
@smlng: I'd say squash, let Murdock run once more and than it could be merged from my point of view.

**Comment by smlng** [comment]
please squash

**Comment by gschorcht** [comment]
Squashed

@maribu @smlng @haukepetersen Just for the sake of completeness. `Wire` is derived public from `Stream`. This PR only implements the methods that are defined in class `Wire`. Theoretically, any other public method of `Stream` could be used in Arduino code. These are:
```
find()
findUntil()
readBytes()
readBytesUntil()
readString()
readStringUntil()
parseInt()
parseFloat()
setTimeout()
```
But the only method from `Stream` I have ever seen for I2C access is `readBytes` which is therefore already implemented in this PR.

We should have this in mind and should implement further functions if there is any request for them. Furthermore, `Serial` is also derived from `Stream`. If there will be further requests for function in future, we should think about to implement a `Stream` class and to derive `Wire` and `Serial` from it.

**Comment by maribu** [comment]
@gschorcht: Thanks for pointing that out.

+1 for not increasing the problem space of this PR. If there really is need for also using the `Stream` methods for I2C, this should be a follow up PR in my opinion. But I personally would not implement that until someone stumbles upon a case it is actually needed. In my opinion it would be a shame to spend effort to add an API that in the end no one ever uses.

**Comment by gschorcht** [comment]
> ACK. If the pointer with the `Stream` API did not raise any new concern, I'd say merge!

As I said, I have never seen problems with the `Stream` class that would block this PR. It was just a comment that it might become necessary in future for enhanced compatibility to think about that class and of course its derivatives.

**Comment by maribu** [comment]
OK, no one complaint. Let's merge!

**Comment by gschorcht** [comment]
@maribu Thanks for your support. Lets's hope we get merged your SPI and the `arduino_lib` build soon.

**Pull request merged**